packages/engine/lib/error/src/report.rs

```diff
@@ -0,0 +1,310 @@
+                frame,
+                #[cfg(feature = "backtrace")]
+                backtrace,
+                #[cfg(feature = "spantrace")]
+                span_trace,
+            }),
+            _context: PhantomData,
+        }
+    }
+
+    /// Creates a new `Report<S>` from a provided scope.
+    #[track_caller]
+    pub fn from_context(context: Context) -> Self
+    where
+        Context: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        #[allow(clippy::option_if_let_else)] // #[track_caller] on closures are unstable
+        let location = if let Some(location) =
+            provider::request_by_type_tag::<'_, tags::FrameLocation, _>(&context)
+        {
+            location
+        } else {
+            Location::caller()
+        };
+
+        #[cfg(feature = "backtrace")]
+        let backtrace =
+            if provider::request_by_type_tag::<'_, tags::ReportBackTrace, _>(&context).is_some() {
+                None
+            } else {
+                Some(Backtrace::capture())
+            };
+
+        #[cfg(feature = "spantrace")]
+        let span_trace =
+            if provider::request_by_type_tag::<'_, tags::ReportSpanTrace, _>(&context).is_some() {
+                None
+            } else {
+                Some(SpanTrace::capture())
+            };
+
+        Self::from_frame(
+            Frame::from_context(context, location, None),
+            #[cfg(feature = "backtrace")]
+            backtrace,
+            #[cfg(feature = "spantrace")]
+            span_trace,
+        )
+    }
+
+    /// Adds a contextual message to the [`Frame`] stack.
+    #[track_caller]
+    pub fn wrap<M>(self, message: M) -> Self
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        Self::from_frame(
+            Frame::from_message(
+                message,
+                Location::caller(),
+                Some(Box::new(self.inner.frame)),
+            ),
+            #[cfg(feature = "backtrace")]
+            self.inner.backtrace,
+            #[cfg(feature = "spantrace")]
+            self.inner.span_trace,
+        )
+    }
+
+    /// Adds context information to the [`Frame`] stack enforcing a typed `Report`.
+    #[track_caller]
+    pub fn provide_context<C>(self, context: C) -> Report<C>
+    where
+        C: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        Report::from_frame(
+            Frame::from_context(
+                context,
+                Location::caller(),
+                Some(Box::new(self.inner.frame)),
+            ),
+            #[cfg(feature = "backtrace")]
+            self.inner.backtrace,
+            #[cfg(feature = "spantrace")]
+            self.inner.span_trace,
+        )
+    }
+
+    /// Converts the `Report<Context>` to `Report<()>` without modifying the frame stack.
+    #[allow(clippy::missing_const_for_fn)] // False positive
+    pub fn compat(self) -> Report {
```

Comment:
I'm a bit lost on the naming here. Maybe we could call it `generalise` or something, if there's a reason that it's related to (I assume) compatibility, can we explain it in the docstring

Reply:
It could be used for compatibility between different modules/crates using `Report`, but I like `generalise` more anyway. ;)